### PR TITLE
Unify Field and ColumnSchema in Storage

### PR DESCRIPTION
**Pull request by imay**
Currently, we have Field and ColumnSchema to access column data in a
row. These two classes are mostly the same. So we should unify these to
one class. Now, Field has offset information, which is an row attribute,
so we remove offset in Field.

RowCursor now has some logic which belong to Schema, so in this patch I
add Schema attribute to RowCursor to make RowCursor simple. After this
change, only Schema will handle Field/ColumnSchema.

I extract some logic from RowCursor to be/src/olap/row.h, then we can
use same logic to handle different types of row. Each type of row has
same function that to get Cell of this row. A cell represent a column
content with a null indicator.

### CODE REVIEW

**Comment by kangkaisen** [comment]
@imay Hi. Why introduce the new concept `ContiguousRow` and `RowCursorCell`? Are they necessary?  I think we should keep core concepts as little as possible and ensure every core concept is clear and easy to understand.

**Comment by imay** [comment]
> @imay Hi. Why introduce the new concept `ContiguousRow` and `RowCursorCell`? Are they necessary? I think we should keep core concepts as little as possible and ensure every core concept is clear and easy to understand.

@kangkaisen 

In origin code, for example in Memtable, we use `char*` to represent a row. I introduce `ContiguousRow` to represent `char* ` with a `Schema`. And I think this struct will replace `RowCursor` in some day. And we have many types of row, like `RowCursor`, `char*`, `RowBlockRow` which is used in storage V2. 

And there are some operations on row like copy, compare with row. In origin code, there are many copies code for same operation. So I introduce `Row` and `Cell` concept to let us finish this operation for all types of row with one copy template row. 

And what's more, in our origin code, many code should handle row format, like `ptr + 1` to get column pointer, `*ptr` to get if it is null. With `Cell` concept, we can encapsulate this logic in it. So If we want to change `RowCursor` format, we need to modify only a couple of files.

File: be/src/olap/schema.h [hunk 1]
```diff
@@ -143,24 +154,30 @@ class Schema {
         *reinterpret_cast<bool*>(row + _col_offsets[index]) = true;
     }
 
-    void set_not_null(int index, char*row) const {
+    void set_not_null(int index, char* row) const {
         *reinterpret_cast<bool*>(row + _col_offsets[index]) = false;
     }
 
+    void set_is_null(void* row, uint32_t cid, bool is_null) const {
```
Comment (chaoyli) [inline be/src/olap/schema.h]: The parameter is not match with set_not_null

File: be/src/olap/wrapper_field.h [hunk 2]
```diff
@@ -83,65 +78,58 @@ class WrapperField {
...
 
     bool is_null() const {
-        return *reinterpret_cast<bool*>(_is_null);
+        return *reinterpret_cast<bool*>(_field_buf);
     }
 
+    void set_is_null(bool is_null) {
+        *reinterpret_cast<bool*>(_field_buf) = is_null;
+    }
     void set_null() {
```
Comment (chaoyli) [inline be/src/olap/wrapper_field.h]: new line

File: be/src/olap/schema.h [hunk 3]
```diff
@@ -81,58 +99,51 @@ class Schema {
...
-        return _col_offsets[index];
+    size_t column_offset(ColumnId cid) const {
+        return _col_offsets[cid];
     }
 
-    size_t get_col_size(int index) const {
-        return _cols[index].size();
+    size_t column_size(ColumnId cid) const {
+        return _cols[cid]->size();
+    }
```
Comment (chaoyli) [inline be/src/olap/schema.h]: new line

File: be/src/olap/row_cursor.h [hunk 4]
```diff
@@ -160,28 +115,44 @@ class RowCursor {
     inline size_t get_fixed_len() const { return _fixed_len; }
     inline size_t get_variable_len() const { return _variable_len; }
 
+    // Get column nullable pointer with column id
+    // TODO(zc): make this return const char*
+    char* nullable_cell_ptr(uint32_t cid) const {
+        return _fixed_buf + _schema->column_offset(cid);
+    }
+    char* cell_ptr(uint32_t cid) const {
```
Comment (chaoyli) [inline be/src/olap/row_cursor.h]: cell_ptr() in RowCursor not include null indicator, in ColumnBlock include null indicator.
The two place are not consistent.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/row_cursor_cell.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+#pragma once
+
+namespace doris {
+
+struct RowCursorCell {
+    RowCursorCell(void* ptr) : _ptr(ptr) { }
+    RowCursorCell(const void* ptr) : _ptr((void*)ptr) { }
+    bool is_null() const { return *reinterpret_cast<bool*>(_ptr); }
+    void set_is_null(bool is_null) const { *reinterpret_cast<bool*>(_ptr) = is_null; }
```
[the remark this answers is not in the document]
Comment (imay) [reply be/src/olap/row_cursor_cell.h]: complier will inline these function even if without inline

File: be/src/olap/row_cursor.h (hunk 4 above)
Comment (imay) [reply be/src/olap/row_cursor.h]: No, cell_ptr in ColumnBlock means doesn't include null indicator. In ColumnBlock, null indicator locates in a bitmap, cell_ptr only return field pointer

File: be/src/olap/schema.h (hunk 1 above)
Comment (imay) [reply be/src/olap/schema.h]: done

**Comment by kangkaisen** [comment]
> > @imay Hi. Why introduce the new concept `ContiguousRow` and `RowCursorCell`? Are they necessary? I think we should keep core concepts as little as possible and ensure every core concept is clear and easy to understand.
> 
> @kangkaisen
> 
> In origin code, for example in Memtable, we use `char*` to represent a row. I introduce `ContiguousRow` to represent `char* ` with a `Schema`. And I think this struct will replace `RowCursor` in some day. And we have many types of row, like `RowCursor`, `char*`, `RowBlockRow` which is used in storage V2.
> 
> And there are some operations on row like copy, compare with row. In origin code, there are many copies code for same operation. So I introduce `Row` and `Cell` concept to let us finish this operation for all types of row with one copy template row.
> 
> And what's more, in our origin code, many code should handle row format, like `ptr + 1` to get column pointer, `*ptr` to get if it is null. With `Cell` concept, we can encapsulate this logic in it. So If we want to change `RowCursor` format, we need to modify only a couple of files.

@imay  OK, Thank you.

**Review by chaoyli — APPROVED** [review]

**Pull request merged**